src/Common/MemoryTracker.cpp

```diff
@@ -205,7 +208,30 @@ void MemoryTracker::allocImpl(Int64 size, bool throw_if_memory_exceeded, MemoryT
         allocation_traced = true;
     }
 
-    if (unlikely(current_hard_limit && will_be > current_hard_limit) && memoryTrackerCanThrow(level, false) && throw_if_memory_exceeded)
+    Int64 limit_to_check = current_hard_limit;
+
+#if USE_JEMALLOC
+    if (level == VariableContext::Global)
+    {
+        /// Jemalloc arenas may keep some extra memory.
+        /// This memory was substucted from RSS to decrease memory drift.
+        /// In case memory is close to limit, try to pugre the arenas.
+        /// This is needed to avoid OOM, because some allocations are directly done with mmap.
+        Int64 current_free_memory_in_allocator_arenas = free_memory_in_allocator_arenas.load(std::memory_order_relaxed);
+
+        if (current_free_memory_in_allocator_arenas > 0 && current_hard_limit && current_free_memory_in_allocator_arenas + will_be > current_hard_limit)
+        {
+            if (free_memory_in_allocator_arenas.exchange(-current_free_memory_in_allocator_arenas))
+            {
+                mallctl("arena." STRINGIFY(MALLCTL_ARENAS_ALL) ".purge", nullptr, nullptr, nullptr, 0);
```

Comment:
Looks like it definitely deserves some metrics (not suggesting logging because it is memory tracker).
What do you think?

_P.S. Now MemoryTracker code looks more and more complicated..._

Reply:
Indeed, should be easy to add.
Now it is more like experimental pr

Reply:
>Now it is more like experimental pr

Well it is merged...